File: service-exif/src/controllers/JobManagerClient.ts
import * as path from 'path';
import * as grpc from '@grpc/grpc-js';
import * as protoloader from '@grpc/proto-loader';
import { ProtoGrpcType } from '../../src/generated/service-jobs';
import { JobManagerControllerClient } from '../../src/generated/jobmanager/JobManagerController';
import config from '../config/server';
import logger, { getLoggerMetaFactory } from '../../../service-shared/logger';
import { GetDataResponse } from '../generated/jobmanager/GetDataResponse';

const PROTO_FILE = '../../protos/service-jobs.proto';

const packageDef = protoloader.loadSync(path.resolve(__dirname, PROTO_FILE));
const grpcObj = grpc.loadPackageDefinition(packageDef) as unknown as ProtoGrpcType;

const makeLogId = getLoggerMetaFactory('JobManagerClient');

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

class JobManagerClient {
    private static client: JobManagerControllerClient;
    private static connectionAttempts = 0;
    static MAX_CONNECTION_ATTEMPTS = 10;
    static TIME_BETWEEN_CONNECTION_ATTEMPTS = 2000;

    static async connect() {
        const logId = makeLogId('connect');
        const timeout = new Date();
        timeout.setSeconds(timeout.getSeconds() + 5);
        const domain = config.grpc.jobManagerHost;
        const port = config.grpc.jobManagerPort;
        const address = `${domain}:${port}`;
        logger.info(`connecting to JobManager gRPC service at: ${address}`);
        this.client = new grpcObj.jobmanager.JobManagerController(
            address,
            grpc.credentials.createInsecure(),
        );

        while (this.connectionAttempts < this.MAX_CONNECTION_ATTEMPTS) {
            this.connectionAttempts += 1;

            logger.info(
                `gRPC connection attempt ${this.connectionAttempts}/${this.MAX_CONNECTION_ATTEMPTS}`,
            );

            const timeout = new Date();
            timeout.setSeconds(timeout.getSeconds() + 5);

            try {
                await new Promise<void>((resolve, reject) => {
                    this.client.waitForReady(timeout, (err) => {
                        if (err) return reject(err);
                        resolve();
                    });
                });

                logger.info(`Connected to JobManager gRPC on ${address}`, logId);
                this.connectionAttempts = 0;
                return;
            } catch (err) {
                logger.warn(`gRPC connection failed: ${err.message}`);
                if (this.connectionAttempts >= this.MAX_CONNECTION_ATTEMPTS) {
                    logger.error('Max connection attempts reached — aborting service.');
                    process.exit(1);
                }

                logger.info(
                    `Retrying in ${this.TIME_BETWEEN_CONNECTION_ATTEMPTS / 1000} seconds...`,
                );
                await sleep(this.TIME_BETWEEN_CONNECTION_ATTEMPTS);
            }
        }
    }

    static getImageData(
        imageSource: string,
        corrId: string,
        jweToken: string,
    ): Promise<Buffer<ArrayBuffer>> {
        const logId = makeLogId('getImageData', corrId);
        return new Promise((resolve, reject) => {
            const meta = new grpc.Metadata();
            meta.set('x-correlation-id', corrId);
            meta.set('authorization', jweToken);

            const stream = this.client.getData({ filepath: imageSource }, meta);
            const buffer: Buffer[] = [];

            stream.on('data', (response: GetDataResponse) => {
                const chunk = response.data as Buffer;
                if (chunk) {
                    buffer.push(chunk);
                }
            });

            stream.on('end', () => {
                const completeBuffer = Buffer.concat(buffer);
                resolve(completeBuffer);
                logger.debug(`completed streaming image data for filepath: ${imageSource}`, logId);
            });

            stream.on('error', (err) => {
                logger.error(`error streaming image data from JobManager: ${err}`, logId);
                reject(err);
            });
        });
    }

    static closeGrpcSocket() {
        this.client.close();
        logger.info('gRPC client closed connection', makeLogId('closeGrpcSocket'));
    }
}

export default JobManagerClient;

File: service_python_shared/service_python_shared/modules/JobManagerClient.py
import asyncio
from typing import Optional, List

import grpc
from grpc import aio

from service_python_shared.configs.config import config
from service_python_shared.generated.service_jobs_pb2 import GetDataRequest
from service_python_shared.generated.service_jobs_pb2_grpc import (
    JobManagerControllerStub,
)
from service_python_shared.modules.logger import get_logger

MAX_CONNECTION_ATTEMPTS = 10
TIME_BETWEEN_ATTEMPTS = 2  # seconds


class JobManagerClient:
    _client: Optional[JobManagerControllerStub] = None
    _channel: Optional[aio.Channel] = None
    _connection_attempts = 0

    @classmethod
    async def connect(cls):
        logger = get_logger("JobManagerClient/connect")
        domain = config["grpc"]["job_manager_host"]
        port = config["grpc"]["job_manager_port"]
        address = f"{domain}:{port}"
        logger.info(f"Connecting to JobManager gRPC service at: {address}")

        cls._channel = aio.insecure_channel(address)
        cls._client = JobManagerControllerStub(cls._channel)

        while cls._connection_attempts < MAX_CONNECTION_ATTEMPTS:
            cls._connection_attempts += 1
            try:
                logger.info(
                    f"Connecting to gRPC attempt {cls._connection_attempts}/{MAX_CONNECTION_ATTEMPTS}"
                )
                await asyncio.wait_for(cls._channel.channel_ready(), timeout=5)
                logger.info(f"Connected to JobManager gRPC on {address}")
                cls._connection_attempts = 0
                return
            except asyncio.TimeoutError:
                logger.error(f"Timed out connecting to JobManager gRPC at {address}")
                if cls._connection_attempts >= MAX_CONNECTION_ATTEMPTS:
                    logger.error("Max connection attempts reached, aborting service!")
                    raise ConnectionError(
                        "Failed to connect to JobManager gRPC service."
                    )
                logger.info(f"Retrying in {TIME_BETWEEN_ATTEMPTS} seconds...")
                await asyncio.sleep(TIME_BETWEEN_ATTEMPTS)

    @classmethod
    async def get_image_data(
        cls, image_source: str, corr_id: str, jwe_token: str
    ) -> bytes:
        if cls._client is None:
            raise RuntimeError("Client not connected. Call connect() first.")

        metadata = (
            ("x-correlation-id", corr_id),
            ("authorization", jwe_token),
        )

        buffer: List[bytes] = []
        log_id = f"getImageData:{corr_id}"

        logger = get_logger("JobManagerClient/get_image_data")

        request = GetDataRequest(filepath=image_source)
        try:
            async for response in cls._client.getData(request, metadata=metadata):
                if response.data:
                    buffer.append(response.data)
        except grpc.RpcError as e:
            logger.error(f"Error streaming image data: {e}", extra={"id": log_id})
            raise

        logger.debug(
            f"Completed streaming image data for {image_source}", extra={"id": log_id}
        )
        return b"".join(buffer)

    @classmethod
    async def close_grpc_socket(cls):
        logger = get_logger("JobManagerClient/close_grpc_socket")
        if cls._channel:
            await cls._channel.close()
            logger.info(
                "gRPC client closed connection", extra={"id": "closeGrpcSocket"}
            )
            cls._client = None
            cls._channel = None
